packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js

```diff
@@ -41,333 +32,320 @@ const DEFAULT_QUERY = {
+			return 0;
+		}
+
+		if ( termASelected && ! termBSelected ) {
+			return -1;
+		}
+
+		if ( ! termASelected && termBSelected ) {
+			return 1;
+		}
+
+		return 0;
+	};
+	termsTree.sort( termOrChildIsSelected );
+	return termsTree;
+}
+
+/**
+ * Find term by parent id or name.
+ *
+ * @param {Object[]}      terms  Array of Terms.
+ * @param {number|string} parent id.
+ * @param {string}        name   Term name.
+ * @return {Object} Term object.
+ */
+function findTerm( terms, parent, name ) {
+	return find( terms, ( term ) => {
+		return (
+			( ( ! term.parent && ! parent ) ||
+				parseInt( term.parent ) === parseInt( parent ) ) &&
+			term.name.toLowerCase() === name.toLowerCase()
+		);
+	} );
+}
+
+/**
+ * Get filter matcher function.
+ *
+ * @param {string} filterValue Filter value.
+ * @return {(function(Object): (Object|boolean))} Matcher function.
+ */
+function getFilterMatcher( filterValue ) {
+	const matchTermsForFilter = ( originalTerm ) => {
+		if ( '' === filterValue ) {
+			return originalTerm;
+		}
+
+		// Shallow clone, because we'll be filtering the term's children and
+		// don't want to modify the original term.
+		const term = { ...originalTerm };
+
+		// Map and filter the children, recursive so we deal with grandchildren
+		// and any deeper levels.
+		if ( term.children.length > 0 ) {
+			term.children = term.children
+				.map( matchTermsForFilter )
+				.filter( ( child ) => child );
+		}
+
+		// If the term's name contains the filterValue, or it has children
+		// (i.e. some child matched at some point in the tree) then return it.
+		if (
+			-1 !==
+				term.name.toLowerCase().indexOf( filterValue.toLowerCase() ) ||
+			term.children.length > 0
+		) {
+			return term;
+		}
+
+		// Otherwise, return false. After mapping, the list of terms will need
+		// to have false values filtered out.
+		return false;
+	};
+	return matchTermsForFilter;
+}
+
+/**
+ * Hierarchical term selector.
+ *
+ * @param {Object} props      Component props.
+ * @param {string} props.slug Taxonomy slug.
+ * @return {WPElement}        Hierarchical term selector component.
+ */
+function HierarchicalTermSelector( { slug } ) {
+	const instanceId = useInstanceId( HierarchicalTermSelector );
+	const [ adding, setAdding ] = useState( false );
+	const [ formName, setFormName ] = useState( '' );
+	/**
+	 * @type {[number|'', Function]}
+	 */
+	const [ formParent, setFormParent ] = useState( '' );
+	const [ showForm, setShowForm ] = useState( false );
+	const [ filterValue, setFilterValue ] = useState( '' );
+	const [ filteredTermsTree, setFilteredTermsTree ] = useState( [] );
+	const debouncedSpeak = useDebounce( speak, 500 );
+
+	const selector = ( select ) => {
+		const { getCurrentPost } = select( editorStore );
+		const { getTaxonomy, getEntityRecords, isResolving } = select(
+			coreStore
+		);
+		const taxonomy = getTaxonomy( slug );
+
+		const availableTerms =
+			getEntityRecords( 'taxonomy', slug, DEFAULT_QUERY ) || [];
```

Comment:
Using the fallback to `[]` would generate a new array on each selector call ideally this constant `EMPTY_ARRAY` should be moved outside of the component/selector (to keep the same instance and avoid rerenders)